File: DSA/01-introduction/solutions.c
#include <stdio.h>
#include <limits.h>
#include <stdlib.h>
#include <time.h>

int find_max(int *nums, int size) {
    if (size == 0) {
        return INT_MIN;  // Return INT_MIN if array is empty
    }

    int max_num = nums[0];  // Initialize with the first element

    for (int i = 1; i < size; i++) {
        if (nums[i] > max_num) {
            max_num = nums[i];
        }
    }

    return max_num;
}

int main(int argc, char *argv[]) {
    int size = 1000;  // Default size, can be modified based on input
    if (argc > 1) {
        sscanf(argv[1], "%d", &size);  // Get size from command-line argument
    }

    // Dynamically allocate memory for the array to handle large input sizes
    int *nums = (int *)malloc(size * sizeof(int));
    if (nums == NULL) {
        printf("Memory allocation failed!\n");
        return 1;
    }

    // Fill array with values 0, 1, 2, ..., size-1
    for (int i = 0; i < size; i++) {
        nums[i] = i;
    }

    clock_t start = clock();
    int max_value = find_max(nums, size);
    clock_t end = clock();

    double time_taken = ((double)(end - start)) / CLOCKS_PER_SEC;

    printf("Max Value: %d\n", max_value);
    printf("Time taken: %f seconds\n", time_taken);

    // Free the allocated memory
    free(nums);

    return 0;
}

File: DSA/01-introduction/01-big-o/solutions.py
import pytest

class Solution():
    def find_max(self, nums: list[int]) -> int:
        """
        Psuedocode:
        1. if nums is empty, return None
        2. enumerate nums
        3. if i == 0, set maxNum = num (first element)
        4. if num > maxNum, set maxNum = num (2 to last element)
        5. return maxNum
        """
        if not nums:
            return None
        for i, num in enumerate(nums):
            if i == 0:
                maxNum = num
            elif num > maxNum:
                maxNum = num
        return maxNum
